script/generate_sql

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+"""
+This script generates SQL files for each SQL file in `sql/` and 
+includes all the temporary UDF definitions used in the queries.
+"""
+
+from argparse import ArgumentParser
+import os, sys
+
+# sys.path needs to be modified to enable package imports from parent
+# and sibling directories. Also see:
+# https://stackoverflow.com/questions/6323860/sibling-package-imports/23542795#23542795
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+from bigquery_etl.parse_udf import parse_udf_dir, udf_usages_in_file
+from textwrap import dedent
+import shutil
+import re
+
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--destination",
+    default="target/sql/",
+    help="The path where generated SQL files will be stored.",
+)
+parser.add_argument(
+    "--udf-dir",
+    default="udf/",
+    help="The directory where declarations of temporary UDFs are stored.",
+)
+parser.add_argument(
+    "--sql-dir",
+    default="sql/",
+    help="The path where files with SQL queries are stored.",
+)
+
+
+def main():
+    args = parser.parse_args()
+
+    parsed_udfs = {x.name: x for x in parse_udf_dir(args.udf_dir)}
+
+    # create output directory if not exists or remove existing files
+    if os.path.exists(args.destination):
+        shutil.rmtree(args.destination)
+    os.makedirs(args.destination, exist_ok=True)
+
+    for root, dirs, sql_files in os.walk(args.sql_dir):
+        for filename in sql_files:
+            # make sure only SQL files are considered
+            if not filename.endswith(".sql"):
+                continue
+
+            # get names of the UDFs used in the SQL queries
+            udf_usages = udf_usages_in_file(os.path.join(root, filename))
+
+            with open(os.path.join(args.destination, filename), "a+") as output_file:
+                # write UDF declarations to file first
+                for udf_usage in udf_usages:
+                    if udf_usage in parsed_udfs:
+                        output_file.write(parsed_udfs[udf_usage].full_sql)
```

Comment:
I think we want `.sql` rather than `.full_sql` here. `full_sql` includes all dependent udf definitions, so we're pulling in the same definition multiple times.

Reply:
Sorry, the raw definitions are in a list `.definitions` rather than a `.sql` string.

Reply:
Thinking more on this, just pasting in `definitions` wouldn't work either, because a query may reference a single UDF without referencing its dependent udfs.

Instead, I think we need to loop through the udf names that we find in the query and pass each one to `parse_udfs.accumulate_dependencies` in order to get each dependent UDF included exactly once.

Reply:
I think `parse_udf.parse_udf_dir` does already take care of accumulating all the dependencies: https://github.com/mozilla/bigquery-etl/blob/62443b1270b08b4384865c49f0a00482cf34db80/tests/parse_udf.py#L71

Reply:
Right, I see what you mean. There are duplicate definitions in the generated files.